{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook is my first attempt at cleaning up the Goyal metadata.\n",
    "\n",
    "I got this metadata from a few email interactions with Andrew Yeh <yeha@upmc.edu>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fexcel = '../../data/raw/goyal2018/FMT_study_log_23Sept2016_edited.xlsx'\n",
    "fprjna = '../../data/raw/goyal2018/goyal2018.PRJNA380944.txt'\n",
    "\n",
    "# Out file\n",
    "fout = '../../data/clean/goyal2018.metadata.txt'\n",
    "\n",
    "outcomes = pd.read_excel(fexcel, sheet_name=\"PUCAI-PCDAI\", skiprows=2)\n",
    "outcomes = outcomes.rename(columns={'Unnamed: 5': 'Notes'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: there are lots of other types of metadata available like medications, adverse events, calprotectin, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient # dds 1/26/15</th>\n",
       "      <th>Screen</th>\n",
       "      <th>Week 1</th>\n",
       "      <th>Month 1</th>\n",
       "      <th>Month 6</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>001 JFC</td>\n",
       "      <td>55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>002 A-Z</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>003 SMA</td>\n",
       "      <td>70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>004 MPM</td>\n",
       "      <td>12.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>005 J-C</td>\n",
       "      <td>55</td>\n",
       "      <td>55.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>35</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Patient # dds 1/26/15 Screen  Week 1  Month 1 Month 6 Notes\n",
       "0               001 JFC     55     NaN      NaN     NaN   NaN\n",
       "1               002 A-Z     50     NaN      NaN     NaN   NaN\n",
       "2               003 SMA     70     NaN      NaN     NaN   NaN\n",
       "3               004 MPM   12.5     0.0     15.0       5   NaN\n",
       "4               005 J-C     55    55.0     35.0      35   NaN"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outcomes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This metadata file includes patients who weren't in the study (which is encoded through colors in the cells...)\n",
    "\n",
    "Let's get the patient IDs out of the ENA metadata.\n",
    "\n",
    "### Clean up ENA metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>submission_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>instrument_platform</th>\n",
       "      <th>...</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sra_aspera</th>\n",
       "      <th>sra_galaxy</th>\n",
       "      <th>cram_index_ftp</th>\n",
       "      <th>cram_index_aspera</th>\n",
       "      <th>cram_index_galaxy</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>broker_name</th>\n",
       "      <th>nominal_sdev</th>\n",
       "      <th>first_created</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA380944</td>\n",
       "      <td>SRP102742</td>\n",
       "      <td>SAMN06652281</td>\n",
       "      <td>SRS2088440</td>\n",
       "      <td>SRX2691263</td>\n",
       "      <td>SRR5396454</td>\n",
       "      <td>SRA550516</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR539/004/SRR5396454</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR539/004/SRR539...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR539/004/SRR5396454</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FMT.03.042.W</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-04-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA380944</td>\n",
       "      <td>SRP102742</td>\n",
       "      <td>SAMN06652280</td>\n",
       "      <td>SRS2088438</td>\n",
       "      <td>SRX2691264</td>\n",
       "      <td>SRR5396455</td>\n",
       "      <td>SRA550516</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR539/005/SRR5396455</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR539/005/SRR539...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR539/005/SRR5396455</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FMT.03.042.P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-04-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA380944</td>\n",
       "      <td>SRP102742</td>\n",
       "      <td>SAMN06652279</td>\n",
       "      <td>SRS2088439</td>\n",
       "      <td>SRX2691265</td>\n",
       "      <td>SRR5396456</td>\n",
       "      <td>SRA550516</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR539/006/SRR5396456</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR539/006/SRR539...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR539/006/SRR5396456</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FMT.03.042.D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-04-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA380944</td>\n",
       "      <td>SRP102742</td>\n",
       "      <td>SAMN06652278</td>\n",
       "      <td>SRS2088441</td>\n",
       "      <td>SRX2691266</td>\n",
       "      <td>SRR5396457</td>\n",
       "      <td>SRA550516</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR539/007/SRR5396457</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR539/007/SRR539...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR539/007/SRR5396457</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FMT.03.040.M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-04-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA380944</td>\n",
       "      <td>SRP102742</td>\n",
       "      <td>SAMN06652277</td>\n",
       "      <td>SRS2088442</td>\n",
       "      <td>SRX2691267</td>\n",
       "      <td>SRR5396458</td>\n",
       "      <td>SRA550516</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR539/008/SRR5396458</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR539/008/SRR539...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR539/008/SRR5396458</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FMT.03.040.W</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-04-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession secondary_study_accession sample_accession  \\\n",
       "0     PRJNA380944                 SRP102742     SAMN06652281   \n",
       "1     PRJNA380944                 SRP102742     SAMN06652280   \n",
       "2     PRJNA380944                 SRP102742     SAMN06652279   \n",
       "3     PRJNA380944                 SRP102742     SAMN06652278   \n",
       "4     PRJNA380944                 SRP102742     SAMN06652277   \n",
       "\n",
       "  secondary_sample_accession experiment_accession run_accession  \\\n",
       "0                 SRS2088440           SRX2691263    SRR5396454   \n",
       "1                 SRS2088438           SRX2691264    SRR5396455   \n",
       "2                 SRS2088439           SRX2691265    SRR5396456   \n",
       "3                 SRS2088441           SRX2691266    SRR5396457   \n",
       "4                 SRS2088442           SRX2691267    SRR5396458   \n",
       "\n",
       "  submission_accession  tax_id       scientific_name instrument_platform  \\\n",
       "0            SRA550516  408170  human gut metagenome            ILLUMINA   \n",
       "1            SRA550516  408170  human gut metagenome            ILLUMINA   \n",
       "2            SRA550516  408170  human gut metagenome            ILLUMINA   \n",
       "3            SRA550516  408170  human gut metagenome            ILLUMINA   \n",
       "4            SRA550516  408170  human gut metagenome            ILLUMINA   \n",
       "\n",
       "       ...                                                 sra_ftp  \\\n",
       "0      ...        ftp.sra.ebi.ac.uk/vol1/srr/SRR539/004/SRR5396454   \n",
       "1      ...        ftp.sra.ebi.ac.uk/vol1/srr/SRR539/005/SRR5396455   \n",
       "2      ...        ftp.sra.ebi.ac.uk/vol1/srr/SRR539/006/SRR5396456   \n",
       "3      ...        ftp.sra.ebi.ac.uk/vol1/srr/SRR539/007/SRR5396457   \n",
       "4      ...        ftp.sra.ebi.ac.uk/vol1/srr/SRR539/008/SRR5396458   \n",
       "\n",
       "                                          sra_aspera  \\\n",
       "0  fasp.sra.ebi.ac.uk:/vol1/srr/SRR539/004/SRR539...   \n",
       "1  fasp.sra.ebi.ac.uk:/vol1/srr/SRR539/005/SRR539...   \n",
       "2  fasp.sra.ebi.ac.uk:/vol1/srr/SRR539/006/SRR539...   \n",
       "3  fasp.sra.ebi.ac.uk:/vol1/srr/SRR539/007/SRR539...   \n",
       "4  fasp.sra.ebi.ac.uk:/vol1/srr/SRR539/008/SRR539...   \n",
       "\n",
       "                                         sra_galaxy  cram_index_ftp  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR539/004/SRR5396454             NaN   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/srr/SRR539/005/SRR5396455             NaN   \n",
       "2  ftp.sra.ebi.ac.uk/vol1/srr/SRR539/006/SRR5396456             NaN   \n",
       "3  ftp.sra.ebi.ac.uk/vol1/srr/SRR539/007/SRR5396457             NaN   \n",
       "4  ftp.sra.ebi.ac.uk/vol1/srr/SRR539/008/SRR5396458             NaN   \n",
       "\n",
       "  cram_index_aspera cram_index_galaxy  sample_alias  broker_name  \\\n",
       "0               NaN               NaN  FMT.03.042.W          NaN   \n",
       "1               NaN               NaN  FMT.03.042.P          NaN   \n",
       "2               NaN               NaN  FMT.03.042.D          NaN   \n",
       "3               NaN               NaN  FMT.03.040.M          NaN   \n",
       "4               NaN               NaN  FMT.03.040.W          NaN   \n",
       "\n",
       "   nominal_sdev  first_created  \n",
       "0           NaN     2017-04-01  \n",
       "1           NaN     2017-04-01  \n",
       "2           NaN     2017-04-01  \n",
       "3           NaN     2017-04-01  \n",
       "4           NaN     2017-04-01  \n",
       "\n",
       "[5 rows x 50 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(fprjna, sep='\\t')\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FMT</th>\n",
       "      <th>sequencing_run</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sample_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FMT</td>\n",
       "      <td>03</td>\n",
       "      <td>042</td>\n",
       "      <td>W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FMT</td>\n",
       "      <td>03</td>\n",
       "      <td>042</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FMT</td>\n",
       "      <td>03</td>\n",
       "      <td>042</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FMT</td>\n",
       "      <td>03</td>\n",
       "      <td>040</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FMT</td>\n",
       "      <td>03</td>\n",
       "      <td>040</td>\n",
       "      <td>W</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   FMT sequencing_run patient_id sample_type\n",
       "0  FMT             03        042           W\n",
       "1  FMT             03        042           P\n",
       "2  FMT             03        042           D\n",
       "3  FMT             03        040           M\n",
       "4  FMT             03        040           W"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "smpls = meta['sample_alias'].str.split('.', expand=True)\n",
    "smpls = smpls.rename(columns={0: 'FMT', \n",
    "                              1: 'sequencing_run', \n",
    "                              2: 'patient_id', \n",
    "                              3: 'sample_type'})\n",
    "smpls.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "meta = pd.merge(meta, smpls, left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's parse the sample types. From Andrew's email:\n",
    "\n",
    "> D refers to the donor sample. When there is a 1 or 2 that means we sent the donor sample for repeat sequencing (most likely because it failed). P is the patient sample before FMT. 4M, 6M, 9M refer to the 4 month, 6 month, and 9 month sample. W is the 1 week sample. WX, P1, P2, PA, PB, 6MX are just samples that underwent repeat sequencing because they failed.\n",
    "\n",
    "Not usre what just `M` is, but the paper includes a 1 month sample so let's guess that it's that. (Also note: the paper doesn't actually look at 4 month or 9 month samples, I wonder why...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['W', 'P', 'D', 'M', '6M', '4M', 'PB', 'D2', 'D1', 'PX', '6MX', 'P1',\n",
       "       'WX', 'P2', 'PA', '9M'], dtype=object)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta['sample_type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "time_dict = {'W': '1_week', \n",
    "             'P': 'pre_fmt',\n",
    "             'D': 'donor', \n",
    "             'M': '1_month',\n",
    "             '4M': '4_month',\n",
    "             '6M': '6_month',\n",
    "             'PB': 'pre_fmt',\n",
    "             'D1': 'donor',\n",
    "             'D2': 'donor',\n",
    "             'PX': 'pre_fmt', \n",
    "             '6MX': '6_month', \n",
    "             'P1': 'pre_fmt', \n",
    "             'WX': '1_week',\n",
    "             'P2': 'pre_fmt',\n",
    "             'PA': 'pre_fmt', \n",
    "             '9M': '9_month'}\n",
    "\n",
    "meta['time_point'] = meta['sample_type'].apply(lambda x: time_dict[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([u'study_accession', u'secondary_study_accession', u'sample_accession',\n",
       "       u'secondary_sample_accession', u'experiment_accession',\n",
       "       u'run_accession', u'submission_accession', u'tax_id',\n",
       "       u'scientific_name', u'instrument_platform', u'instrument_model',\n",
       "       u'library_name', u'library_layout', u'nominal_length',\n",
       "       u'library_strategy', u'library_source', u'library_selection',\n",
       "       u'read_count', u'base_count', u'center_name', u'first_public',\n",
       "       u'last_updated', u'experiment_title', u'study_title', u'study_alias',\n",
       "       u'experiment_alias', u'run_alias', u'fastq_bytes', u'fastq_md5',\n",
       "       u'fastq_ftp', u'fastq_aspera', u'fastq_galaxy', u'submitted_bytes',\n",
       "       u'submitted_md5', u'submitted_ftp', u'submitted_aspera',\n",
       "       u'submitted_galaxy', u'submitted_format', u'sra_bytes', u'sra_md5',\n",
       "       u'sra_ftp', u'sra_aspera', u'sra_galaxy', u'cram_index_ftp',\n",
       "       u'cram_index_aspera', u'cram_index_galaxy', u'sample_alias',\n",
       "       u'broker_name', u'nominal_sdev', u'first_created', u'FMT',\n",
       "       u'sequencing_run', u'patient_id', u'sample_type', u'time_point'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>time_point</th>\n",
       "      <th>n_samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>007</td>\n",
       "      <td>pre_fmt</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>029</td>\n",
       "      <td>donor</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>022</td>\n",
       "      <td>1_week</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>005</td>\n",
       "      <td>6_month</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>010</td>\n",
       "      <td>pre_fmt</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>024</td>\n",
       "      <td>pre_fmt</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>024</td>\n",
       "      <td>donor</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>024</td>\n",
       "      <td>6_month</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>032</td>\n",
       "      <td>6_month</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>033</td>\n",
       "      <td>pre_fmt</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   patient_id time_point  n_samples\n",
       "14        007    pre_fmt          2\n",
       "52        029      donor          2\n",
       "31        022     1_week          2\n",
       "7         005    6_month          2\n",
       "19        010    pre_fmt          2\n",
       "43        024    pre_fmt          2\n",
       "42        024      donor          2\n",
       "41        024    6_month          2\n",
       "66        032    6_month          1\n",
       "73        033    pre_fmt          1"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(meta\n",
    " .groupby(['patient_id', 'time_point'])\n",
    " .size()\n",
    " .reset_index(name='n_samples')\n",
    " .sort_values(by='n_samples', ascending=False)\n",
    ").head(10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, so some of the patients who have duplicate samples sent for sequencing also still have the original. When I look at the data, I should see which sample looks better and only use that one."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clean up clinical data\n",
    "\n",
    "Specifically, extract the patient IDs so I can merge on that with the ENA metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "outcomes['patient_id'] = outcomes['Patient # dds 1/26/15'].str.split(expand=True)[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add responder/remission flags, based on the clinical outcomes.\n",
    "\n",
    "The paper says:\n",
    "\n",
    "> Response was defined as a decrease of 15 points in PUCAI or 12.5 points in PCDAI at 1 month, as used in previous studies. Remission was defined as normalization of previously elevated fecal biomarkers and a PCDAI/PUCAI of 0 points. If subjects required escalation of medical therapy prior to 1-month evaluation, they were considered to be nonresponders. Subsequently, any escalation of medical therapy was considered a loss of response.\n",
    "\n",
    "I don't really know how to figure out who required escalation of medical therapy or something, but I'm assuming that would be reflected in the PCDAI/PUCAI scores..? Anyway, for now let's just go from those.\n",
    "\n",
    "Note: I think that the numbers highlighted in red are the PCDAI scores and the black ones are PUCAI. There aren't very many red ones, so let's just go with everything being PUCAI..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "outcomes.loc[23, 'Notes'] = outcomes.loc[23, 'Month 6']\n",
    "outcomes.loc[23, 'Month 6'] = np.nan"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll define response as _either_ remission (value is 0) OR decrease in at least 15 points."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient # dds 1/26/15</th>\n",
       "      <th>Screen</th>\n",
       "      <th>Week 1</th>\n",
       "      <th>Month 1</th>\n",
       "      <th>Month 6</th>\n",
       "      <th>Notes</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>remission_m1</th>\n",
       "      <th>response_m1</th>\n",
       "      <th>remission_m6</th>\n",
       "      <th>response_m6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>040 RJ</td>\n",
       "      <td>27.5</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>040</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>037 TSH</td>\n",
       "      <td>10</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>037</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>034 RNH</td>\n",
       "      <td>25</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>034</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>011 A-E</td>\n",
       "      <td>0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>011</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>030 CJH</td>\n",
       "      <td>30</td>\n",
       "      <td>32.5</td>\n",
       "      <td>17.5</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>030</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>022 M-M</td>\n",
       "      <td>50</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>5</td>\n",
       "      <td>Week 1 note:  activity affected by flu</td>\n",
       "      <td>022</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>004 MPM</td>\n",
       "      <td>12.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>004</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>029 LPS</td>\n",
       "      <td>55</td>\n",
       "      <td>10.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>029</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>028 AER</td>\n",
       "      <td>30</td>\n",
       "      <td>50.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>028</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>016 M-B</td>\n",
       "      <td>30</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>016</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Patient # dds 1/26/15 Screen  Week 1  Month 1 Month 6  \\\n",
       "37                040 RJ   27.5    10.0     10.0       0   \n",
       "34               037 TSH     10     5.0      5.0       0   \n",
       "32               034 RNH     25     5.0      5.0       0   \n",
       "10               011 A-E      0     5.0      0.0       0   \n",
       "28               030 CJH     30    32.5     17.5       5   \n",
       "21               022 M-M     50    15.0     15.0       5   \n",
       "3                004 MPM   12.5     0.0     15.0       5   \n",
       "27               029 LPS     55    10.0     35.0       5   \n",
       "26               028 AER     30    50.0     15.0      10   \n",
       "15               016 M-B     30     5.0      5.0      15   \n",
       "\n",
       "                                     Notes patient_id  remission_m1  \\\n",
       "37                                     NaN        040         False   \n",
       "34                                     NaN        037         False   \n",
       "32                                     NaN        034         False   \n",
       "10                                     NaN        011          True   \n",
       "28                                     NaN        030         False   \n",
       "21  Week 1 note:  activity affected by flu        022         False   \n",
       "3                                      NaN        004         False   \n",
       "27                                     NaN        029         False   \n",
       "26                                     NaN        028         False   \n",
       "15                                     NaN        016         False   \n",
       "\n",
       "    response_m1  remission_m6  response_m6  \n",
       "37         True          True         True  \n",
       "34        False          True         True  \n",
       "32         True          True         True  \n",
       "10         True          True         True  \n",
       "28        False         False         True  \n",
       "21         True         False         True  \n",
       "3         False         False        False  \n",
       "27         True         False         True  \n",
       "26         True         False         True  \n",
       "15         True         False         True  "
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outcomes['remission_m1'] = outcomes['Month 1'] == 0\n",
    "outcomes['response_m1'] = ((outcomes['Screen'] - outcomes['Month 1']) >= 15) | outcomes['remission_m1']\n",
    "\n",
    "outcomes['remission_m6'] = outcomes['Month 6'] == 0\n",
    "outcomes['response_m6'] = ((outcomes['Screen'] - outcomes['Month 6']) >= 15) | outcomes['remission_m6']\n",
    "\n",
    "outcomes.sort_values(by='Month 6').head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hm ok, not super sure about patient 004 - i.e. not sure whether the authors considered this person a responder or not. I don't expect these data to be that useful anyway, so let's not bother the authors again..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge clinical and ENA metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fullmeta = pd.merge(meta, outcomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fullmeta.to_csv(fout, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
