{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook cleans up the Jacob 2017 metadata file(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# This file was acquired via email communication, and should not be public\n",
    "foutcomes = '../../data/raw/jacob2017/FMT Clinical Response Remission.xlsx'\n",
    "# This file was downloaded from the ENA. \n",
    "fmeta = '../../data/raw/jacob2017/jacob2017.PRJNA388210.txt'\n",
    "\n",
    "fout = '../../data/clean/jacob2017.metadata.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "outcomes = pd.read_excel(foutcomes)\n",
    "meta = pd.read_csv(fmeta, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>submission_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>instrument_platform</th>\n",
       "      <th>...</th>\n",
       "      <th>sra_aspera</th>\n",
       "      <th>sra_galaxy</th>\n",
       "      <th>cram_index_ftp</th>\n",
       "      <th>cram_index_aspera</th>\n",
       "      <th>cram_index_galaxy</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>broker_name</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>nominal_sdev</th>\n",
       "      <th>first_created</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173004</td>\n",
       "      <td>SRS2240022</td>\n",
       "      <td>SRX2869649</td>\n",
       "      <td>SRR5629088</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/008/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/008/SRR5629088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.001.Donor.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.001.Donor.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173005</td>\n",
       "      <td>SRS2240021</td>\n",
       "      <td>SRX2869648</td>\n",
       "      <td>SRR5629089</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/009/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/009/SRR5629089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.007.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.007.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173020</td>\n",
       "      <td>SRS2240020</td>\n",
       "      <td>SRX2869647</td>\n",
       "      <td>SRR5629090</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/000/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/000/SRR5629090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.002.Donor.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.002.Donor.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173021</td>\n",
       "      <td>SRS2240019</td>\n",
       "      <td>SRX2869646</td>\n",
       "      <td>SRR5629091</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/001/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/001/SRR5629091</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.005.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.005.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173018</td>\n",
       "      <td>SRS2240018</td>\n",
       "      <td>SRX2869645</td>\n",
       "      <td>SRR5629092</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/002/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/002/SRR5629092</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.003.W4.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.003.W4.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession secondary_study_accession sample_accession  \\\n",
       "0     PRJNA388210                 SRP108284     SAMN07173004   \n",
       "1     PRJNA388210                 SRP108284     SAMN07173005   \n",
       "2     PRJNA388210                 SRP108284     SAMN07173020   \n",
       "3     PRJNA388210                 SRP108284     SAMN07173021   \n",
       "4     PRJNA388210                 SRP108284     SAMN07173018   \n",
       "\n",
       "  secondary_sample_accession experiment_accession run_accession  \\\n",
       "0                 SRS2240022           SRX2869649    SRR5629088   \n",
       "1                 SRS2240021           SRX2869648    SRR5629089   \n",
       "2                 SRS2240020           SRX2869647    SRR5629090   \n",
       "3                 SRS2240019           SRX2869646    SRR5629091   \n",
       "4                 SRS2240018           SRX2869645    SRR5629092   \n",
       "\n",
       "  submission_accession  tax_id       scientific_name instrument_platform  \\\n",
       "0            SRA567729  408170  human gut metagenome            ILLUMINA   \n",
       "1            SRA567729  408170  human gut metagenome            ILLUMINA   \n",
       "2            SRA567729  408170  human gut metagenome            ILLUMINA   \n",
       "3            SRA567729  408170  human gut metagenome            ILLUMINA   \n",
       "4            SRA567729  408170  human gut metagenome            ILLUMINA   \n",
       "\n",
       "       ...                                              sra_aspera  \\\n",
       "0      ...       fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/008/SRR562...   \n",
       "1      ...       fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/009/SRR562...   \n",
       "2      ...       fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/000/SRR562...   \n",
       "3      ...       fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/001/SRR562...   \n",
       "4      ...       fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/002/SRR562...   \n",
       "\n",
       "                                         sra_galaxy  cram_index_ftp  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/008/SRR5629088             NaN   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/009/SRR5629089             NaN   \n",
       "2  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/000/SRR5629090             NaN   \n",
       "3  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/001/SRR5629091             NaN   \n",
       "4  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/002/SRR5629092             NaN   \n",
       "\n",
       "  cram_index_aspera cram_index_galaxy  \\\n",
       "0               NaN               NaN   \n",
       "1               NaN               NaN   \n",
       "2               NaN               NaN   \n",
       "3               NaN               NaN   \n",
       "4               NaN               NaN   \n",
       "\n",
       "                               sample_alias broker_name  \\\n",
       "0  Longman.FMT.001.Donor.Pool253.Integrated         NaN   \n",
       "1     Longman.FMT.007.W2.Pool253.Integrated         NaN   \n",
       "2  Longman.FMT.002.Donor.Pool253.Integrated         NaN   \n",
       "3     Longman.FMT.005.W2.Pool253.Integrated         NaN   \n",
       "4     Longman.FMT.003.W4.Pool253.Integrated         NaN   \n",
       "\n",
       "                               sample_title  nominal_sdev first_created  \n",
       "0  Longman.FMT.001.Donor.Pool253.Integrated           NaN    2017-11-09  \n",
       "1     Longman.FMT.007.W2.Pool253.Integrated           NaN    2017-11-09  \n",
       "2  Longman.FMT.002.Donor.Pool253.Integrated           NaN    2017-11-09  \n",
       "3     Longman.FMT.005.W2.Pool253.Integrated           NaN    2017-11-09  \n",
       "4     Longman.FMT.003.W4.Pool253.Integrated           NaN    2017-11-09  \n",
       "\n",
       "[5 rows x 51 columns]"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([u'study_accession', u'secondary_study_accession', u'sample_accession',\n",
       "       u'secondary_sample_accession', u'experiment_accession',\n",
       "       u'run_accession', u'submission_accession', u'tax_id',\n",
       "       u'scientific_name', u'instrument_platform', u'instrument_model',\n",
       "       u'library_name', u'nominal_length', u'library_layout',\n",
       "       u'library_strategy', u'library_source', u'library_selection',\n",
       "       u'read_count', u'base_count', u'center_name', u'first_public',\n",
       "       u'last_updated', u'experiment_title', u'study_title', u'study_alias',\n",
       "       u'experiment_alias', u'run_alias', u'fastq_bytes', u'fastq_md5',\n",
       "       u'fastq_ftp', u'fastq_aspera', u'fastq_galaxy', u'submitted_bytes',\n",
       "       u'submitted_md5', u'submitted_ftp', u'submitted_aspera',\n",
       "       u'submitted_galaxy', u'submitted_format', u'sra_bytes', u'sra_md5',\n",
       "       u'sra_ftp', u'sra_aspera', u'sra_galaxy', u'cram_index_ftp',\n",
       "       u'cram_index_aspera', u'cram_index_galaxy', u'sample_alias',\n",
       "       u'broker_name', u'sample_title', u'nominal_sdev', u'first_created'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Longman</td>\n",
       "      <td>FMT</td>\n",
       "      <td>001</td>\n",
       "      <td>Donor</td>\n",
       "      <td>Pool253</td>\n",
       "      <td>Integrated</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Longman</td>\n",
       "      <td>FMT</td>\n",
       "      <td>007</td>\n",
       "      <td>W2</td>\n",
       "      <td>Pool253</td>\n",
       "      <td>Integrated</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Longman</td>\n",
       "      <td>FMT</td>\n",
       "      <td>002</td>\n",
       "      <td>Donor</td>\n",
       "      <td>Pool253</td>\n",
       "      <td>Integrated</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Longman</td>\n",
       "      <td>FMT</td>\n",
       "      <td>005</td>\n",
       "      <td>W2</td>\n",
       "      <td>Pool253</td>\n",
       "      <td>Integrated</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Longman</td>\n",
       "      <td>FMT</td>\n",
       "      <td>003</td>\n",
       "      <td>W4</td>\n",
       "      <td>Pool253</td>\n",
       "      <td>Integrated</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         0    1    2      3        4           5     6     7     8\n",
       "0  Longman  FMT  001  Donor  Pool253  Integrated  None  None  None\n",
       "1  Longman  FMT  007     W2  Pool253  Integrated  None  None  None\n",
       "2  Longman  FMT  002  Donor  Pool253  Integrated  None  None  None\n",
       "3  Longman  FMT  005     W2  Pool253  Integrated  None  None  None\n",
       "4  Longman  FMT  003     W4  Pool253  Integrated  None  None  None"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Expand the sample ID to different columns\n",
    "sampledf = meta['sample_alias'].str.split('.', expand=True)\n",
    "sampledf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          0    1    2    3      4  5   6   7         8\n",
      "43  Longman  FMT  031  Pre  Stool  6  24  16  Project8\n",
      "\n",
      "          0    1    2    3      4  5   6   7         8\n",
      "43  Longman  FMT  031  Pre  Stool  6  24  16  Project8\n",
      "\n",
      "          0    1    2    3      4  5   6   7         8\n",
      "43  Longman  FMT  031  Pre  Stool  6  24  16  Project8\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in [6, 7, 8]:\n",
    "    print(sampledf.dropna(subset=[i]))\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, so one of the samples has some additional fields filled out. Not sure why...?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Longman</td>\n",
       "      <td>FMT</td>\n",
       "      <td>031</td>\n",
       "      <td>Pre</td>\n",
       "      <td>Stool</td>\n",
       "      <td>6</td>\n",
       "      <td>24</td>\n",
       "      <td>16</td>\n",
       "      <td>Project8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Longman</td>\n",
       "      <td>FMT</td>\n",
       "      <td>031</td>\n",
       "      <td>Donor</td>\n",
       "      <td>Project8</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Longman</td>\n",
       "      <td>FMT</td>\n",
       "      <td>031</td>\n",
       "      <td>W4</td>\n",
       "      <td>stool</td>\n",
       "      <td>Project8</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          0    1    2      3         4         5     6     7         8\n",
       "43  Longman  FMT  031    Pre     Stool         6    24    16  Project8\n",
       "61  Longman  FMT  031  Donor  Project8      None  None  None      None\n",
       "76  Longman  FMT  031     W4     stool  Project8  None  None      None"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampledf[sampledf[2] == \"031\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at the other columns that seem less straightforward..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Pool253', 'Project8', 'stool', 'Stool'], dtype=object)"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampledf[4].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Integrated', None, 'Project8', '6'], dtype=object)"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampledf[5].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hm, not sure what these columns are supposed to indicate. Let's just assume that the authors were just inconsistent in their sample labeling, and assume that all samples are the same kind of samples (stool, processed with same method).\n",
    "\n",
    "Next, let's just double check the columns that seem most informative:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['FMT'], dtype=object)"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampledf[1].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['001', '007', '002', '005', '003', '010', '004', '011', '024',\n",
       "       '027', '028', '006', '029', '033', '021', '034', '025', '031',\n",
       "       '012', '019'], dtype=object)"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This column has patient IDs\n",
    "sampledf[2].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Donor', 'W2', 'W4', 'Pre', 'donor'], dtype=object)"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This one is time point (or donor)\n",
    "sampledf[3].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Replace 'Donor' with 'donor'\n",
    "sampledf = sampledf.replace('Donor', 'donor')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's combine this back with the metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>FMT</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Longman</td>\n",
       "      <td>FMT</td>\n",
       "      <td>001</td>\n",
       "      <td>donor</td>\n",
       "      <td>Pool253</td>\n",
       "      <td>Integrated</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Longman</td>\n",
       "      <td>FMT</td>\n",
       "      <td>007</td>\n",
       "      <td>W2</td>\n",
       "      <td>Pool253</td>\n",
       "      <td>Integrated</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Longman</td>\n",
       "      <td>FMT</td>\n",
       "      <td>002</td>\n",
       "      <td>donor</td>\n",
       "      <td>Pool253</td>\n",
       "      <td>Integrated</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Longman</td>\n",
       "      <td>FMT</td>\n",
       "      <td>005</td>\n",
       "      <td>W2</td>\n",
       "      <td>Pool253</td>\n",
       "      <td>Integrated</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Longman</td>\n",
       "      <td>FMT</td>\n",
       "      <td>003</td>\n",
       "      <td>W4</td>\n",
       "      <td>Pool253</td>\n",
       "      <td>Integrated</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_id  FMT patient_id sample_id        4           5     6     7     8\n",
       "0  Longman  FMT        001     donor  Pool253  Integrated  None  None  None\n",
       "1  Longman  FMT        007        W2  Pool253  Integrated  None  None  None\n",
       "2  Longman  FMT        002     donor  Pool253  Integrated  None  None  None\n",
       "3  Longman  FMT        005        W2  Pool253  Integrated  None  None  None\n",
       "4  Longman  FMT        003        W4  Pool253  Integrated  None  None  None"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampledf = sampledf.rename(\n",
    "    columns={0: 'study_id',\n",
    "             1: 'FMT', \n",
    "             2:'patient_id',\n",
    "             3: 'sample_id'})\n",
    "sampledf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "allmeta = pd.merge(\n",
    "    sampledf[['patient_id', 'sample_id']], meta, \n",
    "    left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>submission_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>...</th>\n",
       "      <th>sra_aspera</th>\n",
       "      <th>sra_galaxy</th>\n",
       "      <th>cram_index_ftp</th>\n",
       "      <th>cram_index_aspera</th>\n",
       "      <th>cram_index_galaxy</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>broker_name</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>nominal_sdev</th>\n",
       "      <th>first_created</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>001</td>\n",
       "      <td>donor</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173004</td>\n",
       "      <td>SRS2240022</td>\n",
       "      <td>SRX2869649</td>\n",
       "      <td>SRR5629088</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/008/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/008/SRR5629088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.001.Donor.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.001.Donor.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>007</td>\n",
       "      <td>W2</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173005</td>\n",
       "      <td>SRS2240021</td>\n",
       "      <td>SRX2869648</td>\n",
       "      <td>SRR5629089</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/009/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/009/SRR5629089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.007.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.007.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>002</td>\n",
       "      <td>donor</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173020</td>\n",
       "      <td>SRS2240020</td>\n",
       "      <td>SRX2869647</td>\n",
       "      <td>SRR5629090</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/000/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/000/SRR5629090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.002.Donor.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.002.Donor.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>005</td>\n",
       "      <td>W2</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173021</td>\n",
       "      <td>SRS2240019</td>\n",
       "      <td>SRX2869646</td>\n",
       "      <td>SRR5629091</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/001/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/001/SRR5629091</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.005.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.005.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>003</td>\n",
       "      <td>W4</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173018</td>\n",
       "      <td>SRS2240018</td>\n",
       "      <td>SRX2869645</td>\n",
       "      <td>SRR5629092</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/002/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/002/SRR5629092</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.003.W4.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.003.W4.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  patient_id sample_id study_accession secondary_study_accession  \\\n",
       "0        001     donor     PRJNA388210                 SRP108284   \n",
       "1        007        W2     PRJNA388210                 SRP108284   \n",
       "2        002     donor     PRJNA388210                 SRP108284   \n",
       "3        005        W2     PRJNA388210                 SRP108284   \n",
       "4        003        W4     PRJNA388210                 SRP108284   \n",
       "\n",
       "  sample_accession secondary_sample_accession experiment_accession  \\\n",
       "0     SAMN07173004                 SRS2240022           SRX2869649   \n",
       "1     SAMN07173005                 SRS2240021           SRX2869648   \n",
       "2     SAMN07173020                 SRS2240020           SRX2869647   \n",
       "3     SAMN07173021                 SRS2240019           SRX2869646   \n",
       "4     SAMN07173018                 SRS2240018           SRX2869645   \n",
       "\n",
       "  run_accession submission_accession  tax_id      ...        \\\n",
       "0    SRR5629088            SRA567729  408170      ...         \n",
       "1    SRR5629089            SRA567729  408170      ...         \n",
       "2    SRR5629090            SRA567729  408170      ...         \n",
       "3    SRR5629091            SRA567729  408170      ...         \n",
       "4    SRR5629092            SRA567729  408170      ...         \n",
       "\n",
       "                                          sra_aspera  \\\n",
       "0  fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/008/SRR562...   \n",
       "1  fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/009/SRR562...   \n",
       "2  fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/000/SRR562...   \n",
       "3  fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/001/SRR562...   \n",
       "4  fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/002/SRR562...   \n",
       "\n",
       "                                         sra_galaxy cram_index_ftp  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/008/SRR5629088            NaN   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/009/SRR5629089            NaN   \n",
       "2  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/000/SRR5629090            NaN   \n",
       "3  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/001/SRR5629091            NaN   \n",
       "4  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/002/SRR5629092            NaN   \n",
       "\n",
       "  cram_index_aspera  cram_index_galaxy  \\\n",
       "0               NaN                NaN   \n",
       "1               NaN                NaN   \n",
       "2               NaN                NaN   \n",
       "3               NaN                NaN   \n",
       "4               NaN                NaN   \n",
       "\n",
       "                               sample_alias broker_name  \\\n",
       "0  Longman.FMT.001.Donor.Pool253.Integrated         NaN   \n",
       "1     Longman.FMT.007.W2.Pool253.Integrated         NaN   \n",
       "2  Longman.FMT.002.Donor.Pool253.Integrated         NaN   \n",
       "3     Longman.FMT.005.W2.Pool253.Integrated         NaN   \n",
       "4     Longman.FMT.003.W4.Pool253.Integrated         NaN   \n",
       "\n",
       "                               sample_title nominal_sdev  first_created  \n",
       "0  Longman.FMT.001.Donor.Pool253.Integrated          NaN     2017-11-09  \n",
       "1     Longman.FMT.007.W2.Pool253.Integrated          NaN     2017-11-09  \n",
       "2  Longman.FMT.002.Donor.Pool253.Integrated          NaN     2017-11-09  \n",
       "3     Longman.FMT.005.W2.Pool253.Integrated          NaN     2017-11-09  \n",
       "4     Longman.FMT.003.W4.Pool253.Integrated          NaN     2017-11-09  \n",
       "\n",
       "[5 rows x 53 columns]"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allmeta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>Remission W4</th>\n",
       "      <th>Response W4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>10</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>11</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>12</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>19</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>21</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>24</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>25</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>27</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>28</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>29</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>31</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>33</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>34</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Patient Remission W4  Response W4 \n",
       "0         1            No           No\n",
       "1         2            No           No\n",
       "2         3            No           No\n",
       "3         4            No          Yes\n",
       "4         5            No           No\n",
       "5         6            No           No\n",
       "6         7            No           No\n",
       "7        10            No           No\n",
       "8        11            No           No\n",
       "9        12           Yes          Yes\n",
       "10       19           Yes           No\n",
       "11       21            No           No\n",
       "12       24            No          Yes\n",
       "13       25           Yes          Yes\n",
       "14       27            No          Yes\n",
       "15       28            No          Yes\n",
       "16       29            No           No\n",
       "17       31            No           No\n",
       "18       33            No          Yes\n",
       "19       34            No           No"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outcomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Change the patient ID to an integer in allmeta, so that it matches the outcome data (which is an Excel file, so was automatically converted to an int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "allmeta['patient_id'] = allmeta['patient_id'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "outcomes = outcomes.rename(\n",
    "    columns={'Remission W4 ': 'remission_w4',\n",
    "             'Response W4 ': 'response_w4'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>submission_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>...</th>\n",
       "      <th>cram_index_aspera</th>\n",
       "      <th>cram_index_galaxy</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>broker_name</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>nominal_sdev</th>\n",
       "      <th>first_created</th>\n",
       "      <th>Patient</th>\n",
       "      <th>remission_w4</th>\n",
       "      <th>response_w4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173004</td>\n",
       "      <td>SRS2240022</td>\n",
       "      <td>SRX2869649</td>\n",
       "      <td>SRR5629088</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.001.Donor.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.001.Donor.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>W2</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07172996</td>\n",
       "      <td>SRS2240014</td>\n",
       "      <td>SRX2869641</td>\n",
       "      <td>SRR5629096</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.001.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.001.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>W4</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173013</td>\n",
       "      <td>SRS2239975</td>\n",
       "      <td>SRX2869602</td>\n",
       "      <td>SRR5629135</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.001.W4.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.001.W4.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>W2</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173005</td>\n",
       "      <td>SRS2240021</td>\n",
       "      <td>SRX2869648</td>\n",
       "      <td>SRR5629089</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.007.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.007.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "      <td>7</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>W4</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173016</td>\n",
       "      <td>SRS2240016</td>\n",
       "      <td>SRX2869643</td>\n",
       "      <td>SRR5629094</td>\n",
       "      <td>SRA567729</td>\n",
       "      <td>408170</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.007.W4.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.007.W4.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "      <td>7</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   patient_id sample_id study_accession secondary_study_accession  \\\n",
       "0           1     donor     PRJNA388210                 SRP108284   \n",
       "1           1        W2     PRJNA388210                 SRP108284   \n",
       "2           1        W4     PRJNA388210                 SRP108284   \n",
       "3           7        W2     PRJNA388210                 SRP108284   \n",
       "4           7        W4     PRJNA388210                 SRP108284   \n",
       "\n",
       "  sample_accession secondary_sample_accession experiment_accession  \\\n",
       "0     SAMN07173004                 SRS2240022           SRX2869649   \n",
       "1     SAMN07172996                 SRS2240014           SRX2869641   \n",
       "2     SAMN07173013                 SRS2239975           SRX2869602   \n",
       "3     SAMN07173005                 SRS2240021           SRX2869648   \n",
       "4     SAMN07173016                 SRS2240016           SRX2869643   \n",
       "\n",
       "  run_accession submission_accession  tax_id     ...      cram_index_aspera  \\\n",
       "0    SRR5629088            SRA567729  408170     ...                    NaN   \n",
       "1    SRR5629096            SRA567729  408170     ...                    NaN   \n",
       "2    SRR5629135            SRA567729  408170     ...                    NaN   \n",
       "3    SRR5629089            SRA567729  408170     ...                    NaN   \n",
       "4    SRR5629094            SRA567729  408170     ...                    NaN   \n",
       "\n",
       "  cram_index_galaxy                              sample_alias broker_name  \\\n",
       "0               NaN  Longman.FMT.001.Donor.Pool253.Integrated         NaN   \n",
       "1               NaN     Longman.FMT.001.W2.Pool253.Integrated         NaN   \n",
       "2               NaN     Longman.FMT.001.W4.Pool253.Integrated         NaN   \n",
       "3               NaN     Longman.FMT.007.W2.Pool253.Integrated         NaN   \n",
       "4               NaN     Longman.FMT.007.W4.Pool253.Integrated         NaN   \n",
       "\n",
       "                               sample_title nominal_sdev first_created  \\\n",
       "0  Longman.FMT.001.Donor.Pool253.Integrated          NaN    2017-11-09   \n",
       "1     Longman.FMT.001.W2.Pool253.Integrated          NaN    2017-11-09   \n",
       "2     Longman.FMT.001.W4.Pool253.Integrated          NaN    2017-11-09   \n",
       "3     Longman.FMT.007.W2.Pool253.Integrated          NaN    2017-11-09   \n",
       "4     Longman.FMT.007.W4.Pool253.Integrated          NaN    2017-11-09   \n",
       "\n",
       "  Patient remission_w4  response_w4  \n",
       "0       1           No           No  \n",
       "1       1           No           No  \n",
       "2       1           No           No  \n",
       "3       7           No           No  \n",
       "4       7           No           No  \n",
       "\n",
       "[5 rows x 56 columns]"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allmeta = pd.merge(allmeta, outcomes, left_on='patient_id', right_on='Patient')\n",
    "allmeta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add sample type indicator\n",
    "\n",
    "I need to add a column that indicates whether the sample is a patient or donor sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "allmeta['sample_type'] = allmeta['sample_id'].apply(lambda x: 'donor' if x == 'donor' else 'patient')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finalize metadata and write to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>remission_w4</th>\n",
       "      <th>response_w4</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>...</th>\n",
       "      <th>sra_aspera</th>\n",
       "      <th>sra_galaxy</th>\n",
       "      <th>cram_index_ftp</th>\n",
       "      <th>cram_index_aspera</th>\n",
       "      <th>cram_index_galaxy</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>broker_name</th>\n",
       "      <th>nominal_sdev</th>\n",
       "      <th>first_created</th>\n",
       "      <th>Patient</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR5629088</td>\n",
       "      <td>1</td>\n",
       "      <td>donor</td>\n",
       "      <td>donor</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Longman.FMT.001.Donor.Pool253.Integrated</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173004</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/008/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/008/SRR5629088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.001.Donor.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR5629096</td>\n",
       "      <td>1</td>\n",
       "      <td>W2</td>\n",
       "      <td>patient</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Longman.FMT.001.W2.Pool253.Integrated</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07172996</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/006/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/006/SRR5629096</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.001.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR5629135</td>\n",
       "      <td>1</td>\n",
       "      <td>W4</td>\n",
       "      <td>patient</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Longman.FMT.001.W4.Pool253.Integrated</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173013</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/005/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/005/SRR5629135</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.001.W4.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR5629089</td>\n",
       "      <td>7</td>\n",
       "      <td>W2</td>\n",
       "      <td>patient</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Longman.FMT.007.W2.Pool253.Integrated</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173005</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/009/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/009/SRR5629089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.007.W2.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR5629094</td>\n",
       "      <td>7</td>\n",
       "      <td>W4</td>\n",
       "      <td>patient</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Longman.FMT.007.W4.Pool253.Integrated</td>\n",
       "      <td>PRJNA388210</td>\n",
       "      <td>SRP108284</td>\n",
       "      <td>SAMN07173016</td>\n",
       "      <td>...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/004/SRR562...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/srr/SRR562/004/SRR5629094</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Longman.FMT.007.W4.Pool253.Integrated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-11-09</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession  patient_id sample_id sample_type remission_w4 response_w4  \\\n",
       "0    SRR5629088           1     donor       donor           No          No   \n",
       "1    SRR5629096           1        W2     patient           No          No   \n",
       "2    SRR5629135           1        W4     patient           No          No   \n",
       "3    SRR5629089           7        W2     patient           No          No   \n",
       "4    SRR5629094           7        W4     patient           No          No   \n",
       "\n",
       "                               sample_title study_accession  \\\n",
       "0  Longman.FMT.001.Donor.Pool253.Integrated     PRJNA388210   \n",
       "1     Longman.FMT.001.W2.Pool253.Integrated     PRJNA388210   \n",
       "2     Longman.FMT.001.W4.Pool253.Integrated     PRJNA388210   \n",
       "3     Longman.FMT.007.W2.Pool253.Integrated     PRJNA388210   \n",
       "4     Longman.FMT.007.W4.Pool253.Integrated     PRJNA388210   \n",
       "\n",
       "  secondary_study_accession sample_accession   ...    \\\n",
       "0                 SRP108284     SAMN07173004   ...     \n",
       "1                 SRP108284     SAMN07172996   ...     \n",
       "2                 SRP108284     SAMN07173013   ...     \n",
       "3                 SRP108284     SAMN07173005   ...     \n",
       "4                 SRP108284     SAMN07173016   ...     \n",
       "\n",
       "                                          sra_aspera  \\\n",
       "0  fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/008/SRR562...   \n",
       "1  fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/006/SRR562...   \n",
       "2  fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/005/SRR562...   \n",
       "3  fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/009/SRR562...   \n",
       "4  fasp.sra.ebi.ac.uk:/vol1/srr/SRR562/004/SRR562...   \n",
       "\n",
       "                                         sra_galaxy cram_index_ftp  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/008/SRR5629088            NaN   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/006/SRR5629096            NaN   \n",
       "2  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/005/SRR5629135            NaN   \n",
       "3  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/009/SRR5629089            NaN   \n",
       "4  ftp.sra.ebi.ac.uk/vol1/srr/SRR562/004/SRR5629094            NaN   \n",
       "\n",
       "   cram_index_aspera cram_index_galaxy  \\\n",
       "0                NaN               NaN   \n",
       "1                NaN               NaN   \n",
       "2                NaN               NaN   \n",
       "3                NaN               NaN   \n",
       "4                NaN               NaN   \n",
       "\n",
       "                               sample_alias broker_name nominal_sdev  \\\n",
       "0  Longman.FMT.001.Donor.Pool253.Integrated         NaN          NaN   \n",
       "1     Longman.FMT.001.W2.Pool253.Integrated         NaN          NaN   \n",
       "2     Longman.FMT.001.W4.Pool253.Integrated         NaN          NaN   \n",
       "3     Longman.FMT.007.W2.Pool253.Integrated         NaN          NaN   \n",
       "4     Longman.FMT.007.W4.Pool253.Integrated         NaN          NaN   \n",
       "\n",
       "   first_created Patient  \n",
       "0     2017-11-09       1  \n",
       "1     2017-11-09       1  \n",
       "2     2017-11-09       1  \n",
       "3     2017-11-09       7  \n",
       "4     2017-11-09       7  \n",
       "\n",
       "[5 rows x 57 columns]"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reorder columns for my own sanity\n",
    "first_cols = ['run_accession', 'patient_id', 'sample_id', 'sample_type', 'remission_w4', 'response_w4', 'sample_title']\n",
    "col_order = first_cols + [i for i in allmeta.columns if i not in first_cols]\n",
    "allmeta = allmeta[col_order]\n",
    "allmeta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's do some checks and make sure I didn't duplicate or lose samples in all the merging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "patient_id  sample_id\n",
       "001         W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "002         Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "003         Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "004         Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "005         Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "006         Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "007         Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "010         Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "                        ..\n",
       "024         W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "025         Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "027         Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "028         Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "029         Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "031         Pre          1\n",
       "            W4           1\n",
       "            donor        1\n",
       "033         Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "034         Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "Length: 78, dtype: int64"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(sampledf.groupby(['patient_id', 'sample_id']).size().reset_index()[0].unique())\n",
    "\n",
    "sampledf.groupby(['patient_id', 'sample_id']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "patient_id  sample_id\n",
       "1           W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "2           Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "3           Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "4           Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "5           Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "6           Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "7           Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "10          Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "                        ..\n",
       "24          W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "25          Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "27          Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "28          Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "29          Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "31          Pre          1\n",
       "            W4           1\n",
       "            donor        1\n",
       "33          Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "34          Pre          1\n",
       "            W2           1\n",
       "            W4           1\n",
       "            donor        1\n",
       "Length: 78, dtype: int64"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make sure there aren't duplicate samples\n",
    "print(allmeta.groupby(['patient_id', 'sample_id']).size().reset_index()[0].unique())\n",
    "\n",
    "# Just look at the data\n",
    "allmeta.groupby(['patient_id', 'sample_id']).size()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, looks good. Let's write this to file!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "allmeta.to_csv(fout, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
